src/sign/mod.rs

```diff
@@ -1,10 +1,57 @@
 //! DNSSEC signing.
 //!
 //! **This module is experimental and likely to change significantly.**
+//!
+//! Signatures are at the heart of DNSSEC -- they confirm the authenticity of a
+//! DNS record served by a secure-aware name server.  But name servers are not
+//! usually creating those signatures themselves.  Within a DNS zone, it is the
+//! zone administrator's responsibility to sign zone records (when the record's
+//! time-to-live expires and/or when it changes).  Those signatures are stored
+//! as regular DNS data and automatically served by name servers.
+
 #![cfg(feature = "sign")]
 #![cfg_attr(docsrs, doc(cfg(feature = "sign")))]
 
+use crate::base::iana::SecAlg;
+
+pub mod generic;
 pub mod key;
-//pub mod openssl;
+pub mod openssl;
 pub mod records;
 pub mod ring;
+
+/// Signing DNS records.
+///
+/// Implementors of this trait own a private key and sign DNS records for a zone
+/// with that key.  Signing is a synchronous operation performed on the current
+/// thread; this rules out implementations like HSMs, where I/O communication is
```

Comment:
Given that this trait has `fn sign()` which defines an action, a verb, I would think `Signer` might be a better name for this trait than `Sign`.

Reply:
Also, have you seen https://github.com/NLnetLabs/domain/pull/217/files#diff-af737f2dd1178f239bcc7823d35afed2a2db6947676918377c4e8787c122bf89R28 ?

Reply:
If I'm reading the contributing docs correctly:

> Traits are named using verb phrases that describe what actions the trait represents. An exception are conversion traits that follow the usual Rust convention for naming conversion functions.

That should mean that `Sign` is the appropriate name.  Also see most of the core traits in the standard library like `Copy`, `Clone`, etc.

Reply:
RE PR #217: Most of the time you can get away with `&[u8]`, but I think it's a bit difficult for RR sets (because you might have to reorder them).  I'll replace this with an init-update-finalize interface.

Reply:
Actually, Ed25519 and Ed448 don't support init-update-finalize interfaces -- both Ring and OpenSSL only offer signing functions which take the entire message as a single slice.  Rather than exposing a streaming interface where the underlying implementation performs allocations, I'm going to let the caller organize the data themselves (they may be able to do so more efficiently).

Reply:
> RE PR #217: Most of the time you can get away with `&[u8]`, but I think it's a bit difficult for RR sets (because you might have to reorder them). I'll replace this with an init-update-finalize interface.

I'm a bit confused by this. Surely the caller decides what needs signing, you just sign what you're given? Which re-ordering are you referring to?

Reply:
I meant that init-update-finalize paradigms are nice when the caller has RR data lying in non-contiguous slices, which is almost certainly the case if the RRs have to be sorted first.  That's why callers would likely benefit from having an init-update-finalize interface which lets them avoid allocations.  Unfortunately, EdDSA makes that impossible.